### ISSUE TITLE
bug: failed to transform coproduct with implicit transformer

### ISSUE THREAD

**Issue by andyczerwonka** [scalalandio/chimney#199]
I get an error with what looks to be a pedestrian case.

```scala
import io.scalaland.chimney.Transformer
import io.scalaland.chimney.dsl._
import org.scalatest.funsuite.AnyFunSuite
import org.scalatest.matchers.should.Matchers

sealed trait A extends Product with Serializable
object A {
  case class Foo(s: String) extends A
  case class Bar(kvs: Map[String, Int]) extends A
}

sealed trait B extends Product with Serializable // same structure as A
object B {
  case class Foo(s: String) extends B
  case class Bar(kvs: Map[String, Int]) extends B
}

sealed trait C extends Product with Serializable
object C {
  case class Foo(s: String) extends C
  case class Bar(keys: Seq[String], values: Seq[Int]) extends C
}

class TransformationTest extends AnyFunSuite with Matchers {

  implicit val barTransformer: Transformer[A.Bar, C.Bar] = aBar => {
    val (keys, values) = aBar.kvs.seq.unzip
    C.Bar(keys = keys.toSeq, values = values.toSeq)
  }

  case class Bag[T](xs: Seq[T])

  test("transforming a coproduct with identical structure") {
    val bagA = Bag(Seq(A.Foo("foo"), A.Bar(Map("bar" -> 1))))
    val bagB = bagA.transformInto[Bag[B]]
    succeed
  }

  test("transforming a coproduct with different structure") {
    val bagA = Bag(Seq(A.Foo("foo"), A.Bar(Map("bar" -> 1))))
    val bagC = bagA
      .into[Bag[C]]
      .withFieldComputed(_.xs, _.xs.transformInto[Seq[C]])
      .transform
    succeed
  }

}
```

`A -> B` works, but `A -> C` does not, even if I have a implicit transformer in scope. I get a `Chimney can't derive transformation from Seq[A] to Seq[C]`

![image](https://user-images.githubusercontent.com/138898/102275508-ba0edf80-3ee2-11eb-8f66-881c88365eba.png)

**Comment by andyczerwonka** [comment]
Apparently moving the type definitions outside of the class fixes the problem.

**Comment by andyczerwonka** [comment]
If I write a transformer manually, it works.

```scala
  implicit lazy val aToC: Transformer[A, C] = {
    case A.Foo(s) => C.Foo(s)
    case A.Bar(kvs) =>
      val (keys, values) = kvs.seq.unzip
      C.Bar(keys = keys.toSeq, values = values.toSeq)
  }
```
Which is of course what Chimney lets me avoid, so I must be doing something wrong.

**Comment by andyczerwonka** [comment]
@krzemin just wondering if this issue is validated? Will this be part of an update?

**Comment by krzemin** [comment]
@andyczerwonka thank you for reproduction example :pray: 

It was resolved in #261
